In [84]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### 3A.1. О чём этот модуль

# План модуля:
* Линейная регрессия.

### 3A.2. Линейная регрессия

In [85]:
myData = pd.read_csv('mycar.csv')

In [86]:
myData

Speed  Stopping_dist
0       4              2
1       4             10
2       7              4
3       7             22
4       8             16
5       9             10
6      10             18
7      10             26
8      10             34
9      11             17
10     11             28
11     12             14
12     12             20
13     12             24
14     12             28
15     13             26
16     13             34
17     13             34
18     13             46
19     14             26
20     14             36
21     14             60
22     14             80
23     15             20
24     15             26
25     15             54
26     16             32
27     16             40
28     17             32
29     17             40
30     17             50
31     18             42
32     18             56
33     18             76
34     18             84
35     19             36
36     19             46
37     19             68
38     20             32
39     20             48
40     20             52
41     20             56
42     20             64
43     22             66
44     23             54
45     24             70
46     24             92
47     24             93
48     25             85

In [87]:
X = myData.iloc[:,:-1].values
Y = myData.iloc[:,1].values

In [88]:
from sklearn.model_selection import train_test_split

In [89]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size= 0.3)

In [90]:
from sklearn.linear_model import LinearRegression
myModel = LinearRegression() #Обозначаем, что наша модель - линейная регрессия
myModel.fit(X_train,Y_train) #обучаем модель на обучающих данных

LinearRegression()

In [91]:
y_pred = myModel.predict(X_test)
y_pred

array([32.03121111, 32.03121111,  5.34624828, 15.35310934, 42.03807218,
       45.37369253, 52.04493324, 52.04493324, 28.69559076, 48.70931288,
       55.38055359, 38.70245182, 25.3599704 , 65.38741466, 72.05865537])

### Задание 3.2.1
### Какая из этих задач является задачей линейной регрессии?
Предсказание зарплаты на основе опыта работы, образования и т. д. 

### Задание 3.2.2
### Почему для этих данных не подходит модель линейной регрессии?
Дисперсия шума зависит от x

### Задание 3.2.3
### Какой размер будет у матрицы X (матрица признаков) для данных в таблице? Задача — модель предсказания продаж.
(3*2)

### 3A.3. Ошибки в линейной регрессии

In [92]:
import pandas as pd
import numpy as np

In [93]:
y_happy = [1, 3, 2, 5]

In [94]:
y_happy

[1, 3, 2, 5]

In [95]:
y_happy_pred = [2, 3, -1, 4]

In [96]:
y_happy_pred

[2, 3, -1, 4]

In [97]:
from sklearn import metrics

In [98]:
MAE = metrics.mean_absolute_error(y_happy,y_happy_pred)

print(MAE)

1.25


In [99]:
MSE = metrics.mean_squared_error(y_happy,y_happy_pred)
print(MSE)

2.75


In [100]:
R_2 = metrics.r2_score(y_happy,y_happy_pred)
print(R_2)

-0.2571428571428571


### Задание 3.3.1
### Чему равна MSE на этих данных *F*?
 * 2.75

### Задание 3.3.2
### Что из этого является причиной использования MSE для задачи линейной регрессии?
* Функция потерь непрерывная
* MSE сильнее штрафует за большие ошибки
* MSE выпуклая

### Задание 3.3.3
### У какой из этих моделей большой bias, а у какой — variance?
* У левой большой variance
* У правой большой bias

### 3A.4. Поиск прямой

### Задание 3.4.1
### Для каких данных лучше использовать *OLS*?
* Если в обучающей выборке мало признаков

### Задание 3.4.2

### Задание 3.4.3
* Для первой

### Задание 3.4.4
* Для второй

### Задание 3.4.5
* 0.25

### Задание 3.4.6
* 1

### 3A.5 Линейная регрессия. Предобработка

In [101]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [104]:
import pamdas as pd
import numpy as np 
import matplotlib.pyplot as plt

ModuleNotFoundError: No module named 'pamdas'

In [103]:
data = pd.read_csv('data_flats2.csv', sep =';')
data.head()

FileNotFoundError: [Errno 2] No such file or directory: 'data_flats2.csv'